### DISCUSSION TITLE
Writing a big endian word to spi in viper

### DISCUSSION THREAD

**Discussion by HLammers** [micropython/micropython#15137] [Using MicroPython] [answered]
I have the following function which writes a word-sized register and a word-sized data value to spi, which works perfectly on a Raspberry Pi Pico:
`    def _write_register(self, reg: int, data: int):`
`        dc(0) # set data/command pin low`
`        spi.write(reg.to_bytes(2, 'big'))`
`        dc(1) # set data/command pin high`
`        spi.write(data.to_bytes(2, 'big'))`

To try to gain as much performance boost as possible I‘m trying to translate this into a viper function, but I can’t get it to work.
This doesn’t work:
`    @micropython.viper`
`    def _write_register(self, reg: int, data: int):`
`        dc(0) # set data/command pin low`
`        spi.write(ptr8(reg.to_bytes(2, 'big')))`
`        dc(1) # set data/command pin high`
`        spi.write(ptr8(data.to_bytes(2, 'big')))`

The code hangs on the first spi.write line. I guess that the combination of to_bites and ptr8 must be the problem (I’ve also tried ptr and ptr16, but that doesn’t make any difference).
Can someone explain me what is going wrong here and what I could change to make this work? Is there perhaps a different way to convert a word-sized value to big endian and feed it as bytes to spi.write?

**Comment by robert-hh** [comment]
This short function will hardly get any faster when using viper. Viper is fast at straight python code and pure computations. At  calls to functions and methods it is not faster than byte code or native code, since most of the time is needed for setting up the arguments, performing the call and eventually returning values.

**Reply by HLammers** [reply]
I'm aware of that, but I'm calling this function a lot from loops, so my idea was to copy the code into the loop itself, if I could figure out how to make it work in viper (which would save many times the time it takes to call a function as well)

**Reply by robert-hh** [reply]
You cannot eliminate the call to spi.write(). You may be able to replace reg.to_bytes() into discrete stores into a pre-allocated 2 byte buffer, like:
```
buf=bytearray(2)
...
<inside loop>
  buf[0] = reg >> 8
  buf[1] = reg & 0xff
  spi.write(buf)
```
Maybe that's a little bit faster.

**Reply by HLammers** [reply, marked as answer]
Thanks @robert-hh for your quick suggestion!

I was trying exactly that while you came up with the suggestion :) and this actually works:
`    @micropython.viper`
`    def _write_register(self, reg: int, data: int):`
`        dc(0)`
`        spi.write(bytearray((reg >> 8, reg & 0xFF)))`
`        dc(1)`
`        spi.write(bytearray((data >> 8, data & 0xFF)))`

**Reply by robert-hh** [reply]
That works technically, I only have concern with `bytearray((reg >> 8, reg & 0xFF))`. That causes at least one implicit allocation of memory, maybe two. These may be slower than using a pre-allocated array and direct assignments.

**Reply by HLammers** [reply]
Ah, I see what you mean
